File: src/main/java/me/centralhardware/znatoki/telegram/statistic/report/MonthReport.java
package me.centralhardware.znatoki.telegram.statistic.report;

import me.centralhardware.znatoki.telegram.statistic.eav.Property;
import me.centralhardware.znatoki.telegram.statistic.eav.types.Number;
import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Service;
import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Client;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.PaymentMapper;
import me.centralhardware.znatoki.telegram.statistic.service.ClientService;
import me.centralhardware.znatoki.telegram.statistic.utils.BeanUtils;
import org.apache.commons.lang3.StringUtils;

import javax.ws.rs.core.MultivaluedHashMap;
import java.io.File;
import java.text.Collator;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class MonthReport extends ExcelReport{

    private final List<String> reportFields;
    private final String clientName;
    private final Long userId;

    public MonthReport(String fio,
                       Long userId,
                       Long service,
                       String serviceName,
                       LocalDateTime date,
                       List<String> reportFields,
                       String clientName) {
        super(fio, service, serviceName, date);

        this.reportFields = reportFields;
        this.clientName = clientName;
        this.userId = userId;
    }

    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yy");

    public File generate(List<Service> services){
        services = services
                .stream()
                .filter(it -> it.getServiceId().equals(this.service))
                .toList();
        if (services.isEmpty()) return null;

        newSheet("отчет");

        writeTitle(STR."Отчет по оплате и посещаемости занятий по \{serviceName}", 6);
        writeTitle("Преподаватель: " + fio, 6);

        LocalDateTime dateTime = services.get(0).getDateTime();
        writeTitle(dateTime.format(DateTimeFormatter.ofPattern("MMMM")) + " " +
                dateTime.getYear(), 6);

        List<String> headers = new ArrayList<>();
        headers.add("№");
        headers.add(STR."ФИО \{clientName}");
        BeanUtils.getBean(ClientService.class).findById(services.getFirst().getPupilId())
                .map(Client::getProperties)
                .orElse(Collections.emptyList())
                .stream()
                .map(Property::name)
                .filter(reportFields::contains)
                .forEach(headers::add);
        headers.add("посетил индивидуально");
        headers.add("посетил групповые");
        headers.add("оплата");
        headers.add("Итого");
        headers.add("Даты посещений");

        writeRow(headers.toArray(new String[0]));

        var fioToTimes = new MultivaluedHashMap<Client, Service>();
        services.forEach(it -> BeanUtils.getBean(ClientService.class).findById(it.getPupilId()).ifPresent(client -> fioToTimes.add(client, it)));

        AtomicInteger totalIndividual = new AtomicInteger();
        AtomicInteger totalGroup = new AtomicInteger();

        AtomicInteger i = new AtomicInteger(1);
        var comparator = getComparator(BeanUtils.getBean(ClientService.class).findById(services.getFirst().getPupilId()).get());
        fioToTimes
                .entrySet()
                .stream()
                .sorted(comparator)
                .forEach(it -> {
                    var fioTimes = it.getValue();
                    var client = it.getKey();
                    int individual = (int) fioTimes
                            .stream()
                            .filter(time -> time.getServiceIds().size() == 1)
                            .count();
                    int group = (int) fioTimes
                            .stream()
                            .filter(time -> time.getServiceIds().size() != 1)
                            .count();
                    totalIndividual.addAndGet(individual);
                    totalGroup.addAndGet(group);

                    LinkedHashMap<String, java.lang.Integer> dates = new LinkedHashMap<>();
                    fioTimes.stream().sorted(Comparator.comparing(Service::getDateTime)).forEach(time -> {
                        int timeCount = dates.computeIfAbsent(formatter.format(time.getDateTime()), count -> 0);
                        timeCount++;
                        dates.put(formatter.format(time.getDateTime()), timeCount);
                    });

                    var datesStr = dates.entrySet().stream()
                            .map(entry -> String.format("%s(%s)", entry.getKey(), entry.getValue()))
                            .collect(Collectors.joining(","));

                    List<String> data = new ArrayList<>();
                    data.add(String.valueOf(i.getAndIncrement()));
                    data.add(client.getFio());
                    client.getProperties()
                            .stream()
                            .filter(prop -> reportFields.contains(prop.name()))
                            .map(Property::value)
                            .forEach(data::add);
                    data.add(Integer.toString(individual));
                    data.add(Integer.toString(group));
                    data.add(BeanUtils.getBean(PaymentMapper.class)
                            .getPaymentsSumByPupil(userId, client.getId(), date).toString());
                    data.add("");
                    data.add(datesStr);

                    writeRow(data.toArray(new String[0]));
                });

        writeRow(
                "",
                "Итого",
                "",
                totalIndividual.toString(),
                totalGroup.toString(),
                BeanUtils.getBean(PaymentMapper.class).getPaymentsSum(userId, date).toString()
        );

        return create();
    }

    private Comparator<Map.Entry<Client, List<Service>>> getComparator(Client client){
        Comparator<Map.Entry<Client, List<Service>>> comparator = null;
        var props = client.getProperties()
                .stream()
                .filter(it -> reportFields.contains(it.name()))
                .toList();
        for (var p : props){
            if (comparator == null){
                if (p.type() instanceof Number){
                    comparator = Comparator.comparing(it ->
                            StringUtils.isNotBlank(getProperty(it.getKey(), p.name()).value()) ?
                                    Integer.parseInt(getProperty(it.getKey(), p.name()).value()) :
                                    null,
                            Comparator.nullsLast(Comparator.naturalOrder()));
                } else {
                    comparator = Comparator.comparing(it -> getProperty(it.getKey(), p.name()).value(), Comparator.nullsLast(Comparator.naturalOrder()));
                }
            }
            if (p.type() instanceof Number && StringUtils.isNotBlank(p.value())){
                comparator = comparator.thenComparing(it -> Integer.parseInt(getProperty(it.getKey(), p.name()).value()), Comparator.nullsLast(Comparator.naturalOrder()));
            } else {
                comparator = comparator.thenComparing(it -> getProperty(it.getKey(), p.name()).value(), Comparator.nullsLast(Comparator.naturalOrder()));
            }
        }
        if (comparator == null){
            comparator = Comparator.comparing(it -> it.getKey().getFio(), Comparator.nullsLast(Collator.getInstance(Locale.of("ru", "RU"))));
        } else {
            comparator = comparator.thenComparing(it -> it.getKey().getFio(), Comparator.nullsLast(Collator.getInstance(Locale.of("ru", "RU"))));
        }
        return comparator;
    }

    private Property getProperty(Client client, String name){
        return client.getProperties()
                .stream()
                .filter(it -> it.name().equals(name))
                .findFirst()
                .orElse(null);
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/report/MonthReport.kt
package me.centralhardware.znatoki.telegram.statistic.report

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import com.google.common.collect.Multimaps
import korlibs.time.days
import korlibs.time.hours
import me.centralhardware.znatoki.telegram.statistic.entity.*
import me.centralhardware.znatoki.telegram.statistic.extensions.formatDate
import me.centralhardware.znatoki.telegram.statistic.extensions.formatDateTime
import me.centralhardware.znatoki.telegram.statistic.extensions.print
import me.centralhardware.znatoki.telegram.statistic.mapper.StudentMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.PaymentMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.SubjectMapper
import me.centralhardware.znatoki.telegram.statistic.service.MinioService
import org.apache.poi.ss.usermodel.HorizontalAlignment
import java.io.File
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class MonthReport(
    private val fio: String,
    private val subjectId: SubjectId,
    private val tutorId: TutorId,
) {

    val clients = mutableMapOf<StudentId, Student>()
    fun getStudent(id: StudentId): Student {
        if (!clients.contains(id)) clients[id] = StudentMapper.findById(id)

        return clients[id]!!
    }

    fun generate(times: List<Lesson>, payments: List<Payment>): File? {
        val serviceName = SubjectMapper.getNameById(subjectId)!!

        if (times.isEmpty() && payments.isEmpty()) {
            return null
        }

        val dateTime = times.first().dateTime

        val id2times: Multimap<LessonId, Lesson> =
            Multimaps.synchronizedListMultimap(ArrayListMultimap.create())
        times.forEach { lesson -> id2times.put(lesson.id, lesson) }

        val fioToTimes: Multimap<Student, Lesson> =
            Multimaps.synchronizedListMultimap(ArrayListMultimap.create())
        times.forEach { service ->
            getStudent(service.studentId).let { client ->
                fioToTimes.put(client, service)
            }
        }

        val totalIndividual = AtomicInteger()
        val totalGroup = AtomicInteger()

        val i = AtomicInteger(1)
        val comparator: Comparator<Student> =
            getComparator()

        return excel {
            sheet("отчет") {
                title("Отчет по оплате и посещаемости занятий по $serviceName", 7)
                title("Преподаватель: $fio", 7)
                title(
                    "${dateTime.format(DateTimeFormatter.ofPattern("MMMM", Locale.of("ru")))} ${dateTime.year}",
                    7,
                )
                row {
                    cell("№")
                    cell("ФИО ученика")
                    cell("Класс")
                    cell("посетил индивидуально")
                    cell("посетил групповые")
                    cell("оплата")
                    cell("Итого")
                    cell("Даты посещений")
                }

                fioToTimes
                    .asMap()
                    .toSortedMap(comparator)
                    .forEach { (client, fioTimes) ->
                        val individual = fioTimes.count { id2times[it.id].isIndividual() }
                        val group = fioTimes.count { id2times[it.id].isGroup() }

                        totalIndividual.addAndGet(individual)
                        totalGroup.addAndGet(group)

                        val dates = linkedMapOf<String, Int>()
                        fioTimes
                            .sortedBy { it.dateTime }
                            .forEach { service ->
                                val key = service.dateTime.formatDate()
                                dates[key] = dates.getOrDefault(key, 0) + 1
                            }

                        val datesStr =
                            dates.entries.joinToString(",") { "${it.key}(${it.value})" }

                        row {
                            cell(i.getAndIncrement())
                            cell(client.fio(), HorizontalAlignment.LEFT)
                            cell(client.schoolClass ?: "")
                            cell(individual)
                            cell(group)
                            cell(
                                PaymentMapper.getPaymentsSumForStudent(
                                    tutorId,
                                    fioTimes.first().subjectId,
                                    client.id!!,
                                    dateTime,
                                )
                            )
                            emptyCell()
                            cell(datesStr, HorizontalAlignment.LEFT)
                        }
                    }
                row {
                    emptyCell()
                    cell("Итого")
                    emptyCell()
                    cell(totalIndividual)
                    cell(totalGroup)
                    cell(
                        PaymentMapper.getPaymentsSum(
                            tutorId,
                            times.first().subjectId,
                            dateTime,
                        )
                    )
                }
            }
            sheet("Журнал занятий") {
                title("Журнал занятий по $serviceName", 3)
                title("Преподаватель: $fio", 3)
                title(
                    "${dateTime.format(DateTimeFormatter.ofPattern("MMMM", Locale.of("ru")))} ${dateTime.year}",
                    3,
                )
                row {
                    cell("Дата")
                    cell("ученик")
                    cell("Сумма")
                    cell("Принудительно групповое")
                    cell("Полтора часа")
                    cell("Фото")
                }
                var total = 0.0
                times
                    .sortedByDescending { it.dateTime }
                    .groupBy { it.id }
                    .forEach { id, services ->
                        val fios =
                            if (services.size == 1) {
                                "      " +
                                        getStudent(services.first().studentId).fio()
                            } else {
                                val i = AtomicInteger(1)
                                services.joinToString("\n") {
                                    "${i.getAndAdd(1)} - ${getStudent(it.studentId).fio()}"
                                }
                            }
                        total = total + services.first().amount
                        row {
                            cell(services.first().dateTime.formatDateTime())
                            cell(fios, HorizontalAlignment.LEFT)
                            cell(services.first().amount)
                            if (services.size == 1) {
                                cell(services.first().forceGroup.print())
                            } else {
                                emptyCell()
                            }
                            cell(services.first().extraHalfHour.print())
                            MinioService.getLink(services.first().photoReport!!, 7.days)
                                .onSuccess {
                                    cellHyperlink(it, "отчет")
                                }
                        }
                    }
                row {
                    cell("итого")
                    emptyCell()
                    cell(total)
                }
            }
            sheet("Журнал оплаты") {
                title("Журнал оплаты по $serviceName", 4)
                title("Преподаватель: $fio", 4)
                title(
                    "${dateTime.format(DateTimeFormatter.ofPattern("MMMM", Locale.of("ru")))} ${dateTime.year}",
                    4,
                )
                row {
                    cell("Дата")
                    cell("ученик")
                    cell("Сумма")
                    cell("фото")
                }
                var total = 0
                payments.forEach { payment ->
                    total = total + payment.amount.amount
                    row {
                        cell(payment.dateTime.formatDateTime())
                        cell(getStudent(payment.studentId).fio(), HorizontalAlignment.LEFT)
                        cell(payment.amount)
                        MinioService.getLink(payment.photoReport!!, 3.hours).onSuccess {
                            cellHyperlink(it,"отчет")
                        }
                    }
                }
                row {
                    cell("итого")
                    emptyCell()
                    cell(total)
                }

                title("Примечание: ссылки на отчеты действительны в течение трех часов", 4)

            }
        }
            .build("$fio - $serviceName ${dateTime.format(DateTimeFormatter.ofPattern("MMMM"))} ${dateTime.year}.xlsx")
    }

    private fun getComparator(): Comparator<Student> {
        val comparator: Comparator<Student> = compareBy { it.schoolClass?.value }
        return comparator.thenComparing { it.fio() }
    }


}
